### PR TITLE
GOES16 True Color

### PR DESCRIPTION
**Pull request by julienchastang**
@blaylockbk Just to get the ball rolling here I've issued this PR on your behalf. You have commit author attribution. We'll have to iterate a few times on this PR while @dopplershift and @jrleeman advise us on the proper formatting for the comment sections. Hopefully, the comments can map to markdown to preserve all the formatting you have done.

### CODE REVIEW

**Comment by julienchastang** [comment]
@jrleeman @dopplershift What tool or script are you using the convert the original source `.py` file to other formats including `.ipynb`?

**Comment by dopplershift** [comment]
[sphinx-gallery](https://github.com/sphinx-gallery/sphinx-gallery)

**Comment by julienchastang** [comment]
Related:

https://github.com/mwouts/jupytext
https://news.ycombinator.com/item?id=17950230

**Comment by jrleeman** [comment]
Overall I like the idea of having a "true color" example and this is certainly better documented than what we currently have in the workshop materials. I'm 👎 on basemap though - but would propose moving the example to cartopy.

**Comment by julienchastang** [comment]
My knowledge may be out of date, but last time I looked Basemap had features that cartopy did not. However, in this case, those missing features may not be relevant anyway. For the record, here is [Brian's original notebook](https://github.com/blaylockbk/pyBKB_v3/blob/master/BB_GOES/mapping_GOES16_TrueColor.ipynb).

**Comment by blaylockbk** [comment]
I primarily stuck with Basemap because that's what I know and have used. About every six months, I try to migrate to cartopy, but hit occasional snags that have slowed my adoption. This application, however, I think can be done in cartopy relatively easy. It might take me some time to get around to this again, but definitely willing to make a version that uses cartopy.

**Comment by jrleeman** [comment]
Probably some useful porting resources in the workshop materials: https://github.com/Unidata/python-workshop/blob/master/notebooks/Satellite_Data/Working%20with%20Satellite%20Data.ipynb

**Comment by julienchastang** [comment]
Giving this another shot but with cartopy. Cartopy should, in theory, make things a good deal easier since it handles re-projections more elegantly. Latest push is far from complete (since this is a pretty involved notebook), but is a start and gets to the point of using cartopy. Will keep iterating, slowly...

**Comment by dopplershift** [comment]
I think you can simplify some of the handling of the CF projection metadata using MetPy's xarray helpers, like is done for GINI format data here: https://unidata.github.io/MetPy/latest/examples/formats/GINI_Water_Vapor.html#sphx-glr-examples-formats-gini-water-vapor-py

**Comment by julienchastang** [comment]
I'll check out metpy to grab the meta data, shortly.

With the latest push, I've brought the notebook to the point where we are using cartopy to reproject to Lambert Conformal. @blaylockbk checkout how much simpler the notebook will become once we use cartopy. When we are finally done, it could shrink by a third or more because you won't have to do all the low level stuff you have to do to accommodate Basemap.

**Comment by blaylockbk** [comment]
@julienchastang Nice! I didn't know you could apply a transform for an imshow image like that. It certainly simplifies plotting images with different projections. I like it.

Can Cartopy return a latitude/longitude array for each pixel in the image? That is why I originally did the low-level stuff with `pyproj`. If Cartopy can't return that array, I think the notebook should include the low-level stuff at the end of the example. Having those latitude/longitude values is useful when you want to know specific values at a particular location, even if the lat/lon isn't being used for the plotting transformation.

Forgive me for not knowing how to pull and edit your current document (there is a lot I don't know about GitHub). Maybe we can add this new section at the bottom of the example...
```
# Georeference each pixel
# ==================
# We will use the metadata in the file and `pyproj` to calculate the latitude and longitude of each pixel.
# This is done by using an inverse transformation of the pyproj.Proj map projection object of the map
# coordinates given by X and Y.

# Create a pyproj geostationary map object
p = Proj(proj='geos', h=sat_h, lon_0=sat_lon, sweep=sat_sweep)

# Perform cartographic transformation. That is, convert image projection coordinates (x and y)
# to latitude and longitude values.
XX, YY = np.meshgrid(x, y)
lons, lats = p(XX, YY, inverse=True)

# Because some pixels represent a point in outer space, we assign the pixels not on the globe as an
# arbitrary point in the Gulf of Alaska.
lats[np.isnan(R)] = 57
lons[np.isnan(R)] = -152

# Show a plot to show what we did...
plt.pcolormesh(lons, lats, R)

# Now every GOES ABI pixel has a known latitude and longitude.
print('Latitudes:', lats)
print('Longitudes:', lons)
```
image

**Comment by dopplershift** [comment]
The cartopy projection objects have a `tranform_points` method that you can use to tranform points between coordinate systems. This should work (though untested):
```python
latlon = ccrs.PlateCarree()
geos = ccrs.Geostationary()
lat, lon, _ = latlon.transform_points(geos, x, y)
```

**Comment by julienchastang** [comment]
@dopplershift, referring to the [earlier comment](https://github.com/Unidata/python-gallery/pull/81#issuecomment-466851766), when I try

`
GiniFile(get_test_data('OR_ABI-L2-MCMIPC-M3_G16_s20181781922189_e20181781924562_c20181781925075.nc'))
`

or

`data = xarray.open_dataset(get_test_data('OR_ABI-L2-MCMIPC-M3_G16_s20181781922189_e20181781924562_c20181781925075.nc', False))
`

I get 

`
ValueError: File 'OR_ABI-L2-MCMIPC-M3_G16_s20181781922189_e20181781924562_c20181781925075.nc' is not in the registry.
`

**Comment by dopplershift** [comment]
Right, you should only be using `get_test_data` for data that are part of MetPy's set of testing data. Are you trying to get data from our TDS?

**Comment by julienchastang** [comment]
[Is this what you have in mind](https://github.com/Unidata/python-gallery/pull/81/commits/8913c58a332c87998a465d51f6c1ebeeb2bba738)? I still have to fetch the `perspective_point_height`, I believe.

File: examples/mapping_GOES16_TrueColor.py [hunk 1]
```diff
@@ -269,34 +270,23 @@
...
+dat = C.metpy.parse_cf('CMI_C02')
+geos = dat.metpy.cartopy_crs
 
 # The projection x and y coordinates equals the scanning angle (in radians)
 # multiplied by the satellite height See details here:
 # https://proj4.org/operations/projections/geos.html?highlight=geostationary
+
+sat_h = C['goes_imager_projection'].perspective_point_height
+
 x = C['x'][:] * sat_h
```
Comment (dopplershift) [inline examples/mapping_GOES16_TrueColor.py]: If you use `dat.x` and `dat.y`, MetPy will already have taken care of multiplying them by the satellite height for you. 😉 We try to make our users' lives as easy as we can.

Comment (julienchastang) [reply examples/mapping_GOES16_TrueColor.py]: Out of curiosity, how did you do this? Just tons of if then else logic for various supported projections? Or maybe, the problem is not as unwieldy as I imagine it is.

Comment (dopplershift) [reply examples/mapping_GOES16_TrueColor.py]: In every other CF-compliant projection, the coordinates have a dimensionality of "length", so we can use the unit handling support to change to meters and we're done. We then have a fallback path for the case where the coordinates are in radians and we do the multiplication by the `perspective_point_height`.

Comment (julienchastang) [reply examples/mapping_GOES16_TrueColor.py]: OK, that worked.

**Comment by julienchastang** [comment]
This notebook conversion is asymptotically approaching completion. The following changes have been made with respect to the original notebook:

- [Sample GOES-16 data now served from RAMADDA](http://ramadda-jetstream.unidata.ucar.edu/repository/entry/show?entryid=cb1f0baa-52d8-47a6-b2f6-4a084dc92efc). Therefore data are accessed via DAP protocol. This change should also make the notebook easier to use since the user will not have to fetch the data to get a working example.
- Code now uses Cartopy significantly simplifying notebook.
- Very last cell (after full-disk image) was not ported since I gather it is not important 

Also I had to add `sphinx_markdown_tables` to the `environment.yml`. Travis seems to be complaining about not knowing about `sphinx_markdown_tables`.

@dopplershift @blaylockbk what remains to be done here before an eventual merge? There are a few loose ends I want to take care of on my end like rebasing.

File: environment.yml [hunk 2]
```diff
@@ -12,6 +12,7 @@
    - sphinx
    - sphinx-gallery
    - sphinx_rtd_theme
+   - sphinx_markdown_tables
```
Comment (dopplershift) [inline environment.yml]: sphinx_markdown_tables isn't available from conda-forge. You'll need to tell it to install with pip. Replace this line by adding at the end of the file:
```
    - pip:
        - sphinx-markdown-tables
```

Note the PyPI package name uses dashes, not underscores.

**Comment by blaylockbk** [comment]
I like how this example has been refined. Much simpler, and I learned a few new things, too.

**Comment by julienchastang** [comment]
I think every remaining travis error is for URL lines that are too long (and should not be split IMO). It is also mysteriously complaining about the `metpy` import not being needed even though it is being used. @blaylockbk you have git commit attribution, BTW. I think I have taken this project pretty much as far as I can take it.

**Comment by jrleeman** [comment]
I would fix the line wrapping - nothing wrong with sane wrapping of URLs

**Comment by julienchastang** [comment]
Wrapping URLs in markdown tables and bulleted lists does not work (each for different reason). Can't wrap long URLs embedded in markdown tables. You *can* wrap long url in markdown bulleted lists but Sphinx does not like when translating to `ipynb`. Maybe you know better than me.

Also:  https://stackoverflow.com/questions/10739843/how-should-i-format-a-long-url-in-a-python-comment-and-still-be-pep8-compliant. The point being that it becomes awkward to easily copy paste into the browser.

**Comment by dopplershift** [comment]
I think it would certainly be simpler to just link to @blaylockbk 's GitHub repo, or the folder of notebooks. It would probably lead to people discovering things in that repo better. @julienchastang ?

**Comment by blaylockbk** [comment]
A simple link to the repo or BB_GOES folder would be better than all the hard links. That way if I ever change notebook file names those links wont be broken.

**Comment by julienchastang** [comment]
Latest push will need a rebase before merging. @blaylockbk I linked to the broader https://github.com/blaylockbk/pyBKB_v3/ but maybe that's not what you had in mind and you want to link more narrowly to the GOES stuff. Let me know.

**Comment by blaylockbk** [comment]
That’s ok. It's should be obvious that the GOES stuff is in the GOES folder.

**Comment by julienchastang** [comment]
Woo-hoo! It was kind of a long journey, but we finally got there. Thanks to all those involved.

**Comment by blaylockbk** [comment]
Thanks @julienchastang!

**Pull request merged**